File: adapters/telemetry/tests/test_zeckendorf.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Crystal Warden Supply Chain Labs LLC
"""Zeckendorf/Fibonacci codec — the wire's contract, pinned:
  * exact round-trip over positive integers;
  * doc-faithful small-integer codes (1->11, 2->011, 3->0011, 4->1011);
  * the self-framing invariant: the ONLY '11' in a code is its terminator (so a stream is
    self-delimiting and resynchronizes at the next boundary).
"""
import sys
from pathlib import Path

import pytest

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
import zeckendorf as z  # noqa: E402


# ---------------------------------------------------------------- doc-faithful small codes
@pytest.mark.parametrize("n,code", [
    (1, "11"), (2, "011"), (3, "0011"), (4, "1011"),
    (5, "00011"), (6, "10011"), (7, "01011"), (8, "000011"),
])
def test_small_codes_match_the_doc(n, code):
    assert z.encode(n) == code
    assert z.decode(code) == n


# ---------------------------------------------------------------- round-trip
@pytest.mark.parametrize("n", [1, 2, 3, 12, 99, 100, 1597, 6765, 10_000, 1_000_000])
def test_round_trip_spot(n):
    assert z.decode(z.encode(n)) == n


def test_round_trip_dense_range():
    for n in range(1, 5001):
        assert z.decode(z.encode(n)) == n, f"round-trip failed at {n}"


# ---------------------------------------------------------------- the self-framing invariant
@pytest.mark.parametrize("n", [1, 2, 3, 4, 17, 250, 4181, 99_999])
def test_only_terminator_is_11(n):
    code = z.encode(n)
    assert code.endswith("11"), code
    # the sole occurrence of '11' is the trailing terminator
    assert code.index("11") == len(code) - 2, f"internal '11' in {code!r}"
    # equivalently: the Zeckendorf part (terminator stripped) has no consecutive 1s
    assert "11" not in code[:-1], f"consecutive 1s in Zeckendorf part of {code!r}"


# ---------------------------------------------------------------- stream = concatenation, self-delimited
@pytest.mark.parametrize("values", [
    [1], [1, 1, 1], [4, 2, 7], [1, 4, 1, 100, 3],
    [10_000, 1, 2, 3, 6765], list(range(1, 200)),
])
def test_stream_round_trip(values):
    assert z.decode_stream(z.encode_stream(values)) == values


def test_small_ints_are_tiny():
    # delta-differenced telemetry lives on small ints; they must be short
    assert all(len(z.encode(n)) <= 6 for n in range(1, 9))


# ---------------------------------------------------------------- rejections
def test_rejects_non_positive():
    with pytest.raises(ValueError):
        z.encode(0)
    with pytest.raises(ValueError):
        z.encode(-3)


def test_rejects_non_code():
    with pytest.raises(ValueError):
        z.decode("10")        # does not end in '11'
    with pytest.raises(ValueError):
        z.decode("1")         # too short

File: prismpath-telemetry-rs/tests/test_zeckendorf.rs
use prismpath_telemetry_rs::zeckendorf as z;

#[test]
fn test_small_codes_match_the_doc() {
    let cases = vec![
        (1, "11"),
        (2, "011"),
        (3, "0011"),
        (4, "1011"),
        (5, "00011"),
        (6, "10011"),
        (7, "01011"),
        (8, "000011"),
    ];
    for (n, code) in cases {
        assert_eq!(z::encode(n).unwrap(), code);
        assert_eq!(z::decode(code).unwrap(), n);
    }
}

#[test]
fn test_round_trip_spot() {
    for n in [1, 2, 3, 12, 99, 100, 1597, 6765, 10_000, 1_000_000] {
        assert_eq!(z::decode(&z::encode(n).unwrap()).unwrap(), n);
    }
}

#[test]
fn test_round_trip_dense_range() {
    for n in 1..=5000 {
        let code = z::encode(n).unwrap();
        assert_eq!(z::decode(&code).unwrap(), n, "round-trip failed at {n}");
    }
}

#[test]
fn test_only_terminator_is_11() {
    for n in [1, 2, 3, 4, 17, 250, 4181, 99_999] {
        let code = z::encode(n).unwrap();
        assert!(code.ends_with("11"), "{code}");
        assert_eq!(code.find("11").unwrap(), code.len() - 2, "internal '11' in {code:?}");
        assert!(!code[..code.len() - 1].contains("11"), "consecutive 1s in Zeckendorf part of {code:?}");
    }
}

#[test]
fn test_stream_round_trip() {
    let cases: Vec<Vec<usize>> = vec![
        vec![1],
        vec![1, 1, 1],
        vec![4, 2, 7],
        vec![1, 4, 1, 100, 3],
        vec![10_000, 1, 2, 3, 6765],
        (1..200).collect(),
    ];
    for values in cases {
        assert_eq!(z::decode_stream(&z::encode_stream(&values).unwrap()), values);
    }
}

#[test]
fn test_small_ints_are_tiny() {
    for n in 1..9 {
        assert!(z::encode(n).unwrap().len() <= 6);
    }
}

#[test]
fn test_rejects_non_positive() {
    assert!(z::encode(0).is_err());
}

#[test]
fn test_rejects_non_code() {
    assert!(z::decode("10").is_err());
    assert!(z::decode("1").is_err());
}

#[test]
fn test_encode_stream_rejects_zero() {
    assert!(z::encode_stream(&[3, 0, 5]).is_err());
    assert!(z::encode(0).is_err());
}

#[test]
fn test_decode_stream_strict_accepts_padding_and_rejects_truncation() {
    let bits = z::encode_stream(&[4, 2, 7]).unwrap();
    // clean stream decodes
    assert_eq!(z::decode_stream_strict(&bits).unwrap(), vec![4, 2, 7]);
    // trailing zeros = byte-alignment padding, accepted
    let padded = format!("{bits}0000");
    assert_eq!(z::decode_stream_strict(&padded).unwrap(), vec![4, 2, 7]);
    // a trailing '1' with no terminator = truncated final code, rejected (the lenient
    // decode_stream silently drops it — the exact difference the strict variant exists for)
    let truncated = format!("{bits}010");
    assert_eq!(z::decode_stream(&truncated), vec![4, 2, 7]);
    assert!(z::decode_stream_strict(&truncated).is_err());
}
